File: brace-expansion.js
// Time Complexity : O(K^(N/K)) Where N is the length of the string and K is the number of groupings I have
// Space Complexity : O(N) Where N is the length of the string
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {string} S
 * @return {string[]}
 */
let result;
const expand = (S) => {
    result = [];
    const choiceList = parseList(S);
    backtrack(choiceList, [], 0);
    return result.sort();
};
    
const parseList = (S) => {
    const choiceList = [];
    let index = 0;

    while (index < S.length) {
        let currChoices = [];
        if (S[index] == '{') {
            index++;
            while (S[index] != '}') {
                if (S[index] != ',') currChoices.push(S[index]);
                index++;
            }
        } else {
            currChoices.push(S[index]);
        }
        choiceList.push(currChoices);
        index++;
    }
    
    return choiceList;
}

const backtrack = (choiceList, temp, index) => {
    // Base case
    if (temp.length >= choiceList.length) {
        result.push(temp.join(''));
        return;
    }
    
    // Logic
    for (let i = 0; i < choiceList[index].length; i++) {
        // Action
        temp.push(choiceList[index][i])
        // Recurse
        backtrack(choiceList, temp, index + 1);
        // Backtrack
        temp.pop();
    }
}

File: Brace Expansion.py
class Solution:
    def expand(self, s: str) -> List[str]:
        #Approach: Backtracking
        #Time Complexity: O(n)
        #Space Complexity: O(n * l)
        #where, n is the number of possible words, and l is the average length of a word
        
        self.result = []
        self.backtrack(s, [], 0)
        return self.result
    
    def backtrack(self, s, sb, idx):
        #base
        if idx == len(s):
            self.result.append(''.join(sb))
            return
        
        #logic
        choices = []
        if s[idx].isalpha():
            choices.append(s[idx])
            
        else:   #s[idx] == '{'
            while s[idx] != '}':
                if s[idx].isalpha():
                    choices.append(s[idx])
                idx += 1
            choices.sort()
            
        for choice in choices:
            sb.append(choice)                       # action
            self.backtrack(s, sb, idx + 1)          # recursion
            sb.pop()                                # backtracking